ygen/protoelements.go

```diff
@@ -116,14 +117,94 @@ func (s *genState) yangTypeToProtoScalarType(args resolveTypeArgs) (*mappedType,
 			return nil, fmt.Errorf("cannot map enumeration without context entry: %v", args)
 		}
 		return &mappedType{nativeType: yang.CamelCase(args.contextEntry.Name)}, nil
+	case yang.Yidentityref:
+		if args.contextEntry == nil {
+			return nil, fmt.Errorf("cannot map identityref without context entry: %v", args)
+		}
+		return &mappedType{
+			nativeType: fmt.Sprintf("%s.%s.%s", basePackageName, enumPackageName, s.resolveIdentityRefBaseType(args.contextEntry)),
+		}, nil
+	case yang.Yunion:
+		return s.protoUnionType(args, basePackageName, enumPackageName)
 	default:
 		// TODO(robjs): implement missing types.
-		//	- identityref
 		//	- binary
 		//	- bits
-		//	- union
-		return nil, fmt.Errorf("unimplemented type: %s", args.yangType.Kind)
+		return nil, fmt.Errorf("unimplemented type in scalar generation: %s", args.yangType.Kind)
+	}
+}
+
+// protoUnionType resolves the types that are included within the YangType in resolveTypeArgs into the
+// scalar type that can be included in a protobuf oneof. The basePackageName and enumPackageName are used
+// to determine the paths that are used for enumerated types within the YANG schema. Each union is
+// resolved into a oneof that contains the scalar types, for example:
+//
+// leaf a {
+//	type union {
+//		type string;
+//		type int32;
+//	}
+// }
+//
+// Is represented in the output protobuf as:
+//
+// oneof a {
+//	string a_string = NN;
+//	int32 a_int32 = NN;
+// }
+//
+// The mappedType's unionTypes can be output through a template into the oneof.
+func (s *genState) protoUnionType(args resolveTypeArgs, basePackageName, enumPackageName string) (*mappedType, error) {
+	var errs []error
+	unionTypes := make(map[string]int)
+	// Extract the subtypes that are defined into map, keyed on the union type.
+	for _, subtype := range args.yangType.Type {
```

Comment:
Seems like this for loop could be eliminated by calling protoUnionSubTypes with args.yangType?

Reply:
True, done.